src/Data/Array/Mutable/Unlifted/Linear.hs

```diff
@@ -0,0 +1,107 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE LinearTypes #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE UnboxedTuples #-}
+{-# LANGUAGE UnliftedNewtypes #-}
+
+-- |
+-- This module defines a linear wrapper around 'GHC.MutableArray#'
+-- which can be manipulated in pure contexts.
```

Comment:
We should work on this description. The module defines an unlifted array type, with a pure linear interface, which makes sure that many operations can be implemented by mutation, rather that copy. Something to this effect.

Reply:
Reading it again, you're right. The current documentation just talks about the implementation details. So, what do you think about:

> This modules defines an unlifted array type with a linear interface, where the operations are implemented with in-place updates rather than copying whenever possible.

Reply:
That's still too developer focused, I think. Maybe just:

> This module provides an unlifted mutable array with a pure interface. Though the array itself is unlifted, it's elements are lifted types. This is made possible by using linear types to make sure array references are single threaded through reads and writes.

Reply:
@Divesh-Otwani I like it. Should we swap your second and third sentence, because I think the "this" on your third sentence refers to the first sentence rather than the second?